File: 0221-maximal-square/0221-maximal-square.java
class Solution {
    public int maximalSquare(char[][] matrix) {
        int m = matrix.length;
        int n = matrix[0].length;
        int[][] dp = new int[m+1][n+1];
        
        for(int i=1; i<=m; i++) {
            for(int j=1; j<=n; j++) {
                dp[i][j] = matrix[i-1][j-1] == '1' ? 1 : 0;
            }
        }
        
        for(int i=m; i>=1; i--) {
            for(int j=n; j>=1; j--) {
                if(dp[i][j] > 0) {
                    int top = dp[i-1][j];
                    int left = dp[i][j-1];
                    int diagnol = dp[i-1][j-1];
                    
                    if(top > 0 && left > 0 && diagnol > 0) {
                        int min = Math.min(top, Math.min(left, dp[i][j]));
                        dp[i-1][j-1] = min + 1;
                    }
                }
            }
        }
        
        int max = -1;
        for(int i=1; i<=m; i++) {
            for(int j=1; j<=n; j++) {
                // System.out.print(dp[i][j] + " ");
                max = Math.max(max, dp[i][j]);
            }
            // System.out.println();
        }
        return max*max;
    }
}

File: 0221-maximal-square/0221-maximal-square.js
/**
 * @param {character[][]} matrix
 * @return {number}
 */
var maximalSquare = function(matrix) {
    const m = matrix.length
    const n = matrix[0].length
    const grid = new Array(m+1)
    for(let i=0; i<m+1; i++) {
        grid[i] = new Array(n+1).fill(0)
        for(let j=0; i<m && j<n; j++) {
            grid[i][j] = parseInt(matrix[i][j])
        }
    }
    let max = 0
    for(let i = m-1; i>=0; i--) {
        for(let j=n-1; j>=0; j--) {
            if(grid[i][j] === 1) {
                const min = Math.min(grid[i][j+1], grid[i+1][j], grid[i+1][j+1])
                grid[i][j] = min+1
                max = Math.max(max, grid[i][j])
            }
        }
    }

    // console.log(grid)
    return max*max

};